### PR TITLE
fix(APIv2): RHINENG-2097 allow profiles and rules to be indexed by ref_id

### PR DESCRIPTION
**Pull request by RoamingNoMaD**
## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

File: app/controllers/v2/profiles_controller.rb [hunk 1]
```diff
@@ -20,7 +22,11 @@ def profiles
     end
 
     def profile
-      @profile ||= authorize(expand_resource.find(permitted_params[:id]))
+      @profile ||= if ::UUID.validate(permitted_params[:id])
```
Comment (skateman) [inline app/controllers/v2/profiles_controller.rb:25]: I would put this logic into the concern as it is repetitive in more than one controller.

File: app/controllers/v2/rules_controller.rb [hunk 2]
```diff
@@ -3,6 +3,8 @@
 module V2
   # API for Rules under Security Guides
   class RulesController < ApplicationController
+    include Referring
```
Comment (skateman) [inline app/controllers/v2/rules_controller.rb:6]:
```suggestion
    include IndexableByRefId
```

File: spec/controllers/v2/concerns/rendering_spec.rb [hunk 3]
```diff
@@ -135,6 +135,22 @@
     expect(response.parsed_body).to match(expected)
   end
 
+  context 'referrable resource', if: [V2::ProfilesController, V2::RulesController].include?(described_class) do
```
Comment (skateman) [inline spec/controllers/v2/concerns/rendering_spec.rb:138]: This logic is the opposite of declarative, instead please create a new file called `indexable_by_ref_id_spec.rb` under the concerns and define a shared example there that you explicitly specify.

File: spec/controllers/v2/profiles_controller_spec.rb [hunk 4]
```diff
@@ -46,5 +46,6 @@
     let(:notfound_params) { extra_params.merge(security_guide_id: FactoryBot.create(:v2_security_guide).id) }
 
     it_behaves_like 'individual', :security_guide
+    it_behaves_like 'indexable by ref id', :security_guide
```
Comment (skateman) [inline spec/controllers/v2/profiles_controller_spec.rb:49]:
```suggestion
    it_behaves_like 'indexable by ref_id', :security_guide
```

File: app/controllers/concerns/v2/indexable_by_ref_id.rb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        if ::UUID.validate(id)
+          scope.find(id)
+        else
+          scope.find_by(ref_id: referize(id))
+        end
+      end
+
+      private
+
+      def referize(id)
```
Comment (skateman) [inline app/controllers/concerns/v2/indexable_by_ref_id.rb:19]:
```suggestion
      def to_ref_id(id)
```

File: app/controllers/concerns/v2/indexable_by_ref_id.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module V2
+  # Logic for referring to resources with ref_id
+  module IndexableByRefId
+    extend ActiveSupport::Concern
+
+    included do
+      def find_by_id_or_ref_id(scope, id)
```
Comment (skateman) [inline app/controllers/concerns/v2/indexable_by_ref_id.rb:9]:
```suggestion
      def ref_id_lookup(scope, id)
```

**Review by skateman — CHANGES_REQUESTED** [review]

**Comment by romanblanco** [comment]
Failing QE test does not seem to be relevant:

```
10:38:44 [2023-11-24T10:38:44.553Z] ==================================== ERRORS ====================================
10:38:44 [2023-11-24T10:38:44.553Z] ___________ ERROR at teardown of test_graphql_host_groups_filtering ____________
10:38:44 [2023-11-24T10:38:44.553Z] 
10:38:44 [2023-11-24T10:38:44.553Z]     @request.addfinalizer
10:38:44 [2023-11-24T10:38:44.553Z]     def _finalizer():
10:38:44 [2023-11-24T10:38:44.553Z] >       second_system.delete()
10:38:44 [2023-11-24T10:38:44.553Z] 
10:38:44 [2023-11-24T10:38:44.553Z] /iqe_venv/lib/python3.9/site-packages/iqe_compliance/tests/test_graphql.py:1063:
```

**Review by romanblanco — APPROVED** [review]
LGTM :+1:

**Review by skateman — APPROVED** [review]
The Seal of Approval

**Pull request merged**